### ISSUE TITLE
[Core][Nightly-test] scheduling_test_many_0s_tasks_many_nodes failed

### ISSUE THREAD

**Issue by scv119** [ray-project/ray#24234]
### What happened + What you expected to happen

https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_ZSSKSXJ5pjHAcws5g3Z2eGgM?command-history-section=command_history

```
[2m[33m(raylet, ip=172.31.70.203)[0m [2022-04-24 20:11:54,239 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(2377) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.67.141)[0m [2022-04-24 20:11:54,373 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(2521) have not registered within the timeout. The process is still alive, probably it's hanging during start.
...
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,747 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4031) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,756 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4032) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,761 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4033) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,768 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4034) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,771 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4035) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,775 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4036) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.69.48)[0m [2022-04-24 20:12:47,093 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(3486) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.69.48)[0m [2022-04-24 20:12:47,098 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(3487) have not registered within the timeout. The process is still alive, probably it's hanging during start.
Traceback (most recent call last):
  File "distributed/test_scheduling.py", line 105, in <module>
    args.total_num_actors, args.num_actors_per_nodes, job
  File "distributed/test_scheduling.py", line 53, in start_actor
    lambda: ray.get([actor.do_job.remote() for actor in actors])
  File "distributed/test_scheduling.py", line 38, in measure
    ret = f()
  File "distributed/test_scheduling.py", line 53, in <lambda>
    lambda: ray.get([actor.do_job.remote() for actor in actors])
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 1827, in get
    raise value
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
```

### Versions / Dependencies

lastet ray

### Reproduction script

run nightly test

### Issue Severity

High: It blocks me from completing my task.

**Comment by scv119** [comment]
@jjyao to triage.

**Comment by jjyao** [comment]
It's flaky, it failed 2 times in the past 24 runs.

**Comment by jjyao** [comment]
raylet was killed by oom
```
[ 1003.003069] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=9d0aea09100bdef5543d2a5802a50d5de241edf2e15df64873b2b94ee715d4e5,mems_allowed=0-1,global_oom,task_memcg=/docker/37edd43b2906212c15644dc01d55ff5c51d794c6712bcca736e01adb7d3c384a,task=raylet,pid=7855,uid=1000
[ 1003.003397] Out of memory: Killed process 7855 (raylet) total-vm:323813084kB, anon-rss:244045284kB, file-rss:0kB, shmem-rss:12kB, UID:1000 pgtables:478352kB oom_score_adj:0
[ 1009.613149] oom_reaper: reaped process 7855 (raylet), now anon-rss:0kB, file-rss:0kB, shmem-rss:12kB
```

**Comment by scv119** [comment]
is it a regression? if not new it doesn't necessary need to block release.

**Comment by jjyao** [comment]
It has been flaky since March.

**Comment by jjyao** [comment]
```
ray._private.memory_monitor.RayOutOfMemoryError: More than 95% of the memory on node ip-172-31-91-75 is used (242.54 / 246.38 GB). The top 10 memory consumers are:

PID   MEM     COMMAND
248   225.92GiB       /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet --raylet_socket_name=
1876  6.52GiB ray::SimpleActor
186   2.11GiB /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server --log_dir=/tmp/ray/s
```

**Comment by mwtian** [comment]
Most recent failure: https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_9KjYG8zWGDdApYGr6HhELc8j?command-history-section=command_history
Similar out of memory issue:
```
(SimpleActor pid=1740) ray._private.memory_monitor.RayOutOfMemoryError: More than 95% of the memory on node ip-172-31-68-53 is used (239.66 / 246.38 GB). The top 10 memory consumers are:
(SimpleActor pid=1740) 
(SimpleActor pid=1740) PID      MEM     COMMAND
(SimpleActor pid=1740) 241      222.0GiB        /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet --raylet_socket_name=
(SimpleActor pid=1740) 1740     7.4GiB  ray::SimpleActor
(SimpleActor pid=1740) 186      1.98GiB /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server --log_dir=/tmp/ray/s
(SimpleActor pid=1740) 211      0.26GiB /home/ray/anaconda3/bin/python -u /home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/dash
(SimpleActor pid=1740) 301      0.15GiB /home/ray/anaconda3/bin/python -u /home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/agen
(SimpleActor pid=1740) 77       0.11GiB /home/ray/anaconda3/bin/python /home/ray/anaconda3/bin/anyscale session web_terminal_server --deploy
(SimpleActor pid=1740) 423      0.11GiB /home/ray/anaconda3/bin/python /home/ray/anaconda3/bin/anyscale session auth_start
(SimpleActor pid=1740) 4120     0.07GiB ray::IDLE
(SimpleActor pid=1740) 1780     0.07GiB ray::IDLE
(SimpleActor pid=1740) 1777     0.07GiB ray::IDLE
(SimpleActor pid=1740) 
(SimpleActor pid=1740) In addition, up to 0.0 GiB of shared memory is currently being used by the Ray object store.
```

**Comment by iycheng** [comment]
@mwtian  do you have some updates about this issue?  It never succeeded after 6/02

**Comment by mwtian** [comment]
#25516 may help with debugging a bit.

**Comment by mwtian** [comment]
Still investigating. The behavior where one Raylet uses much more memory (> 200GB) than others (< 20GB) can be reliably reproduced in the nightly test. But It seems with jemalloc instrumentation, this behavior is gone and Raylet no longer OOMs.

**Comment by mwtian** [comment]
Current observations:
1. When using 32 nodes in the test, it is consistent that exactly 1 node has abnormal Raylet memory usage. When using 16 or 24 nodes in the test, no Raylet has abnormal memory usage.

2. Even after the test finishes and there is no longer any workload on the Ray cluster, the node still has the same high memory usage. e.g.
<img width="1913" alt="Screen Shot 2022-06-09 at 11 58 37 AM" src="https://user-images.githubusercontent.com/81660174/172944416-e716d26b-3e01-457d-a134-bba6cd77cfcf.png">

3. When the Raylet starts to have increasing memory usage, it is also starting to spawn workers.

4. The Raylet with high memory usage also receives abnormal number of `CancelWorkerLease` RPCs. The correlation is strong (Raylet with highest memory usage <=> Raylet receiving highest number of `CancelWorkerLease` RPCs) but I'm not sure of the causation here. Memory usage can grow or stay the same independent of the trend of  new `CancelWorkerLease` calls. `CancelWorkerLease` has a suspicious retry logic in the client but it can be a separate issue.
<img width="1913" alt="Screen Shot 2022-06-09 at 11 58 37 AM" src="https://user-images.githubusercontent.com/81660174/172944900-08dda6fc-eafd-437d-9c44-8878d672cb6c.png">

5. The memory usage increase disappears when using `jemalloc`, with or without profiling enabled.

Since memory profiling seems blocked by unable to reproduce the issue with profiling enabled, the next step is to fix all the suspicious behavior found during the investigations (e.g. #25626), and try to reproduce the issue on a single node.

**Comment by scv119** [comment]
Jemalloc: makes problem go away
ASAN: makes problem go away
TSAN: lots of false positives

**Comment by mwtian** [comment]
I'm hopeful that #26112 can mitigate the issue i.e. making the test pass consistently. Downgrading to P1 for now, to decide if we want to find the root cause.

**Comment by nirn1973** [comment]
Got a similar crash on Centos 7 (not on Windows):
186016 UserWarning: Distributing <class 'dict'> object. This may take some time.
 186017 ^[[2m^[[33m(raylet)^[[0m [2022-07-28 03:54:50,260 C 21557 21557] raylet_client.cc:57: Could not connect to socket /tmp/ray/session_2022-07-27_20-33-29_503982_        94220/sockets/raylet
 186018 ^[[2m^[[33m(raylet)^[[0m *** StackTrace Information ***
 186019 ^[[2m^[[33m(raylet)^[[0m     ray::SpdLogMessage::Flush()
 186020 ^[[2m^[[33m(raylet)^[[0m     ray::RayLog::~RayLog()
 186021 ^[[2m^[[33m(raylet)^[[0m     ray::raylet::RayletConnection::RayletConnection()
 186022 ^[[2m^[[33m(raylet)^[[0m     ray::raylet::RayletClient::RayletClient()
 186023 ^[[2m^[[33m(raylet)^[[0m     ray::core::CoreWorker::CoreWorker()
 186024 ^[[2m^[[33m(raylet)^[[0m     ray::core::CoreWorkerProcessImpl::CreateWorker()
 186025 ^[[2m^[[33m(raylet)^[[0m     ray::core::CoreWorkerProcessImpl::CoreWorkerProcessImpl()
 186026 ^[[2m^[[33m(raylet)^[[0m     ray::core::CoreWorkerProcess::Initialize()
 186027 ^[[2m^[[33m(raylet)^[[0m     __pyx_pw_3ray_7_raylet_10CoreWorker_1__cinit__()
 186028 ^[[2m^[[33m(raylet)^[[0m     __pyx_tp_new_3ray_7_raylet_CoreWorker()
 186029 ^[[2m^[[33m(raylet)^[[0m     type_call
 186030 ^[[2m^[[33m(raylet)^[[0m
 186031 ^[[2m^[[33m(raylet)^[[0m [2022-07-28 03:55:06,098 E 94431 94431] worker_pool.cc:481: Some workers of the worker process(21557) have not registered within the         timeout. The process is dead, probably it crashed during start.
